### ISSUE TITLE
Meaning of EPS2, threshold / Parameter tuning for rainfall

### ISSUE THREAD

**Issue by alexandrefierro** [metno/titanlib#20]
Dear Cristian:

Now that I successfully implemented Titanlib into our software, I was wondering if you could (based on your experience) offer some suggestions on how to tune the following input parameters for (e.g., ) hourly rainfall ?

NB: what are the layman's meanings of eps2 and threshold?:

threshold=0.1 
min_std = 1 
min_horizontal_scale=1000
vertical_scale=200 
num_min_prof=0  # I assume  0 = turned off? If not, how to turn this option off?
pos=np.ones(N) * 0.1
neg=np.ones(N) * 0.1
eps2=np.ones(N) * 0.15

Cheers and best,
Alex-

**Comment by cristianlussana** [comment]
Dear Alex,

I can try to help. I assume you have downloaded the master branch. Which test are you implementing? Buddy check?

For the buddy check:
- threshold is (from the wiki) "The buddy check flags observations if the (absolute value of the) difference between the observations and the average of the neighbours normalized by the standard deviation in the circle is greater than a predefined >>threshold<<" 
- eps2. It is used for SCT and not for the buddy check. The SCT compares the observed value at a location with an estimate obtained considering the neighbouring observations, let's call this estimate cross-validation value (cv). eps2 is a parameter that is used to obtain cv by means of statistical interpolation. if eps2 is close to 0 (e.g. 0.1-0.2), then the interpolation will try to fit all surrounding observations. if it is close to 1, then the interpolation will stay closer to the regional mean value of the observations.

How to optimize the parameters, try some values (eps2: 0.1-0.2-0.3...-1), thresholds (0.5-1-2-3...)

Four hourly rainfall and given your settings:
- if this is buddy check, threshold=0.1 seems rather small;
- min_std=1 ok, you may even try larger values
- min_horizontal_scale=1000 m, ok if you have a dense network
- vertical scale=200 m, here I will put 99999999 . large number, just to remove the elevation from the test (usually is not that important for hourly rainfall or you do not have enough observations to sample the near-surface vertical profile properly)
- num_min_prof used for SCT, set it to a negative number, this will turn off this option
- pos/neg, 0.1 is  a very small number, you will remove a lot of observations. Try something larger (1-2-3-4-...), then refine given the results you obtain
- eps2, first guesses: 0.1-0.25-0.5-1. Find the best one and keep it
- 
For rainfall, it might be worth playing also with some kind of transformation before running the quality control. The point is that you may want to transform the data, such that the distribution of their values is more Gaussian-like.

Hope this helps.

Best, Cristian

**Comment by alexandrefierro** [comment]
Hello Cristian!

I have installed the package through the following command, so I assume this refers to the master code bundle:

python -m pip install titanlib==0.3.0

My QC is performed through the following sequence of calls:

flags  = titanlib.range_check(values, minimum, maximum)
flags2 = titanlib.buddy_check()
flags3, sct, rep = titanlib.sct()
flags4 = titanlib.isolation_check()

wherein:

TOTAL_FLAG=flags+flags2+flags3+flags4                                                                   
print(f"Number of flagged Stations in combined Titan checks --> {np.count_nonzero(TOTAL_FLAG)}")

So, in a nutshell, if any of these tests “fails” the station is removed from the control cohort.

The code runs as intended and I obtained the best results with these values below.
I use a std=5.0 because of the large gradients often seen in precipitation fields, especially during the summer:

   pos=np.ones(N) * 5.0                                                 
   neg=np.ones(N) * 5.0                                                           
   eps2=np.ones(N) * 0.5  

   threshold=0.1

Following your suggestions, I will also conduct additional tests on “Gaussian-ized“ rainfall data.

Best and thank you again very much for your time and help; much appreciated.
Alex-
